Client code refined from "Dynamic streams add/delete at run-time #181 #198 #218"

JunsubHan4rmEricsson opened the pull request:
Hi I've made a few changes according to your feedback.
@Cyborn-Minjae-Lee @egwakim @ejangle @JaeyongYu @jsmoon @JunsubHan4rmEricsson @taewoonyun @yonghwan007 @youngjun1

These are the changes that have taken place that are different from #218 
**1) Profile state update**
: async event for profile state needs to be implemented both in server and client side (in addition to the exisitng async port state event)
**2) Explicit "Remove" of a profile**
: --remove function for "stop command" needs to be supported (instead of recycling the reset command)
**3) "Streams" for each profile**
: streams command need to show profile information as well

**In addition, the following regression tests also pass**
test_basic_cont
test_basic_multi_burst
test_basic_single_burst

In other words, duration issues have been fixed.
Please provide feedback.

Thanks.

hhaim commented:
I will not have time to review this today. 
I would start working on the following in parallel to finish this feature. 
1. Add documentation into STL doc 
2. Add system test -- see my proposal for the first test

hhaim commented:
I would cover the flow stats /latency with dynamic profile too

JunsubHan4rmEricsson commented:
Right, I will talk to our tester about it and implement it.
Do you have any other feedback ?

JunsubHan4rmEricsson commented:
Can you be a bit more specific about "native API snippet"
Where should it be added? Client API? Tests?

hhaim commented:
To the the same Stateless_manual.asciidoc and not in the rpc manual.
I would like an easy start for automation users that want to use this feature, just cut and past from a running example in the manual.

You could add this snippet to the python API documentation too, but this is for advanced users

JunsubHan4rmEricsson commented:
1) Documentation 
- 1.a) Use case example of "simple native API snippet" in stateless_manual.asciidoc

2) Tests
- 2.a) Flow stats test
- 2.b) Latency test
- 2.c) Random addition of profile and async event check for each addition (sort of stress test)

Will that be all?
Plus, what are some commands you might use for 1.a. 
Is there a specific set of commands you would like to show as an example?

JunsubHan4rmEricsson commented:
Hi Hanoh,

Among the three requirements, I have completed and merged 1.a) and 2.c).

**For 1.a) I have added a use case from which a user can understand how to use profile functionality.** 
https://github.com/cisco-system-traffic-generator/trex-core/pull/220/commits/8a07fa8c4ca52a040ccf0503be2aff8a3b1c9807

**For 2.c) I have added a test case, "test_basic_cont_dynamic_profile"**
https://github.com/cisco-system-traffic-generator/trex-core/pull/220/commits/1966648dd66faa26b2d7c2549efa33ac36388832

I'm currently working on 2.a) and 2.b) with my colleagues. 
Thanks.


**Documentation**
1.a) Use case example of "simple native API snippet" in stateless_manual.asciidoc

**Tests**
2.a) Flow stats test

- 2.b) Latency test
- 2.c) Random addition of profile and async event check for each addition (sort of stress test)

hhaim commented:
Good. Will try to integrate it shortly. 
BTW, have a look into this  https://github.com/cisco-system-traffic-generator/trex-core/commit/29367cf31426e9f157eb83de9aeac79c15a8ee3d
This is something I missed in the code review and it seems we don't have a test for that

JunsubHan4rmEricsson commented:
1) I will have my colleague, who was in charge of server implementation, take a look at **29367cf**
2) I have added a test case for profile version of flow_stats/latency test **"test_latency_pause_resume_dynamic_profile"**
https://github.com/cisco-system-traffic-generator/trex-core/pull/220/commits/9918822e202c3facfb6a501ea8e498a92d5bbfd2

Please have a look and provide any feedback
Thanks

JunsubHan4rmEricsson commented:
> Good. Will try to integrate it shortly.
> BTW, have a look into this [29367cf](https://github.com/cisco-system-traffic-generator/trex-core/commit/29367cf31426e9f157eb83de9aeac79c15a8ee3d)
> This is something I missed in the code review and it seems we don't have a test for that

For this test case, we have not added anything for GUI environment. 
So I'm not exactly sure what should be done from our side.

Cyborn-Minjae-Lee commented:
Hi Hanoh,

Regarding 29367cf , Thanks so much for your fixing.
As an excuse, I have no way to test and verify for the functionality of GUI.

BR,
Minjae

hhaim commented:
Could you create one branch with both Dynamic_Profile_Client and Profile_Client_Refined code rebased one master. 
There are conflicts betwean the branchs

JunsubHan4rmEricsson commented:
You can forget about [Client Implementation] Dynamic streams add/delete at run-time #181 #198 #218
This branch (Profile_Client_Refined #220 ) is the one with all the necessary changes including the other one (Dynamic_Profile_Client #218 )

hhaim commented:
regarding [diff](https://github.com/cisco-system-traffic-generator/trex-core/pull/220/commits/1966648dd66faa26b2d7c2549efa33ac36388832#diff-afbbd9bd2620278300f771a4111bad5eR752) better to start and stop each profile not at the **same time**  see my proposal

hhaim commented:
> You can forget about [Client Implementation] Dynamic streams add/delete at run-time #181 #198 #218
> This branch (Profile_Client_Refined #220 ) is the one with all the necessary changes including the other one (Dynamic_Profile_Client #218 )

Understood. will try it I saw that it wasn't fork from the same parent

JunsubHan4rmEricsson commented:
> regarding [diff](https://github.com/cisco-system-traffic-generator/trex-core/pull/220/commits/1966648dd66faa26b2d7c2549efa33ac36388832#diff-afbbd9bd2620278300f771a4111bad5eR752) better to start and stop each profile not at the **same time** see my proposal

I'm sorry.
What exactly do you mean?

hhaim commented:
I mean that I had in mind that I need to merge both branchs Profile_Client + _Profile_Client_Refined
will try only this branch.

JunsubHan4rmEricsson commented:
Thanks
Let me know if you have any more feedback

hhaim commented:
In general the events works but it crash with simple senarios 

see this:

```
1. reload server 
2. run this 
trex>stop --remove -p 0.*
Traceback (most recent call last):
  File "/sw/packages/python3/current/lib/python3.4/runpy.py", line 171, in _run_module_as_main
    "__main__", mod_spec)
  File "/sw/packages/python3/current/lib/python3.4/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/auto/srg-sce-swinfra-usr/emb/users/hhaim/work/depot/asr1k/emb/private/hhaim/bp_sim_git/trex-core/scripts/automation/trex_control_plane/interactive/trex/console/trex_console.py", line 937, in <module>
    main()
  File "/auto/srg-sce-swinfra-usr/emb/users/hhaim/work/depot/asr1k/emb/private/hhaim/bp_sim_git/trex-core/scripts/automation/trex_control_plane/interactive/trex/console/trex_console.py", line 926, in main
    console.start()
  File "/auto/srg-sce-swinfra-usr/emb/users/hhaim/work/depot/asr1k/emb/private/hhaim/bp_sim_git/trex-core/scripts/automation/trex_control_plane/interactive/trex/console/trex_console.py", line 643, in start
    self.cmdloop()
  File "/sw/packages/python3/current/lib/python3.4/cmd.py", line 137, in cmdloop
    line = self.precmd(line)
  File "/auto/srg-sce-swinfra-usr/emb/users/hhaim/work/depot/asr1k/emb/private/hhaim/bp_sim_git/trex-core/scripts/automation/trex_control_plane/interactive/trex/console/trex_console.py", line 293, in precmd
    stop = self.onecmd(line)
  File "/sw/packages/python3/current/lib/python3.4/cmd.py", line 217, in onecmd
    return func(arg)
  File "/auto/srg-sce-swinfra-usr/emb/users/hhaim/work/depot/asr1k/emb/private/hhaim/bp_sim_git/trex-core/scripts/automation/trex_control_plane/interactive/trex/common/trex_api_annotators.py", line 125, in wrap2
    rc = f(*args)
  File "/auto/srg-sce-swinfra-usr/emb/users/hhaim/work/depot/asr1k/emb/private/hhaim/bp_sim_git/trex-core/scripts/automation/trex_control_plane/interactive/trex/stl/trex_stl_client.py", line 1993, in stop_line
    opts = parser.parse_args(line.split(), default_ports = self.get_profiles_with_state("active"), verify_acquired = True, allow_empty = True)
  File "/auto/srg-sce-swinfra-usr/emb/users/hhaim/work/depot/asr1k/emb/private/hhaim/bp_sim_git/trex-core/scripts/automation/trex_control_plane/interactive/trex/utils/parsing_opts.py", line 1192, in parse_args
    self.client.psv.validate(self.cmd_name, opts.ports, PSV_ACQUIRED, allow_empty = allow_empty)
  File "/auto/srg-sce-swinfra-usr/emb/users/hhaim/work/depot/asr1k/emb/private/hhaim/bp_sim_git/trex-core/scripts/automation/trex_control_plane/interactive/trex/common/trex_psv.py", line 251, in validate
    if has_dup(ports):
  File "/auto/srg-sce-swinfra-usr/emb/users/hhaim/work/depot/asr1k/emb/private/hhaim/bp_sim_git/trex-core/scripts/automation/trex_control_plane/interactive/trex/utils/common.py", line 254, in has_dup
    return len(l) > len(set(l))
TypeError: unhashable type: 'PortProfileID'
Shutting down RPC client
```

2. 

**Unhandled event 10**
``` 

-=TRex Console v3.0=-

Type 'help' or '?' for supported actions

trex>start -f stl/imix.py -p  0.2 -m 10kpps -d 10

Removing all streams from port(s) [0.2]:                     [SUCCESS]


Attaching 3 streams to port(s) [0.2]:                        [SUCCESS]


Starting traffic on port(s) [0.2]:                           Unhandled event 10
[SUCCESS]

57.26 [ms]

trex>

Unhandled event 10   <<<<<

```

JunsubHan4rmEricsson commented:
-=TRex Console v3.0=-

Type 'help' or '?' for supported actions

trex>stop --remove -p 0.*
no active ports
no ports with streams
534.06 [usec]

When I start server, then client
and run "stop --remove -p 0.*"
it works fine like shown above.

How did you reload server?

hhaim commented:
I've rebased it.
https://github.com/hhaim/trex-core/tree/bidtt4-STL_Dynamic_Profile_Client_Refined

```
sudo ./t-rex-64 -i
```

try to connect/disconnect a few times (using quit)
in the first time it works

hhaim commented:
When you run the regression you should see the ```Unhandled event 11``` message 
Didn't you see this?
```
Unhandled event 10
Done (10s), CPU util: 0.113771, norm_pps_per_core: 36.17mpps norm_bw_per_core: 433.24Gb/core
Testing profile stl/udp_for_benchmarks.py, kwargs: {'packet_len': 4000}
Unhandled event 11
Unhandled event 11
Unhandled event 11
Unhandled event 11
Unhandled event 10
Unhandled event 10
Unhandled event 10
Unhandled event 10
Done (10s), CPU util: 0.0513993, norm_pps_per_core: 30.26mpps norm_bw_per_core: 966.58Gb/core
Testing profile stl/udp_for_benchmarks.py, kwargs: {'packet_len': 9000}
Unhandled event 11
Unhandled event 11
Unhandled event 11
Unhandled event 11
Unhandled event 10
Unhandled event 10
Unhandled event 10
Unhandled event 10
Done (3s), CPU util: 0.034354, norm_pps_per_core: 20.23mpps norm_bw_per_core: 1220.3Gb/core
Testing profile stl/udp_for_benchmarks.py, kwargs: {'stream_count': 10, 'packet_len': 9000}
Unhandled event 11
Unhandled event 11
Unhandled event 11
Unhandled event 11
Unhandled event 10
Unhandled event 10
Unhandled event 10
Unhandled event 10
```

hhaim commented:
regarding the events looking in to the code I see the issue. 


```
class ServerEventsIDs(object):
    """
        server event IDs
        (in sync with the server IDs)
    """
    EVENT_PORT_STARTED   = 0
    EVENT_PORT_STOPPED   = 1
    EVENT_PORT_PAUSED    = 2
    EVENT_PORT_RESUMED   = 3
    EVENT_PORT_JOB_DONE  = 4
    EVENT_PORT_ACQUIRED  = 5
    EVENT_PORT_RELEASED  = 6
    EVENT_PORT_ERROR     = 7
    EVENT_PORT_ATTR_CHG  = 8

    EVENT_PROFILE_STARTED       = 10
    EVENT_PROFILE_STOPPED       = 11
    EVENT_PROFILE_PAUSED        = 12
    EVENT_PROFILE_RESUMED       = 13
    EVENT_PROFILE_FINISHED_TX   = 14
    EVENT_PROFILE_ERROR         = 17

    EVENT_ASTF_STATE_CHG = 50

    EVENT_SERVER_STOPPED = 100





class TRexSubscriber():

```
You should to handle all the events (even the EVENT_PROFILE_STARTED). This is for read-only clients. which should know that the profile X was started and stopped. For the main client there is no need to handle the EVENT_PROFILE_STARTED for example.

As I said try to connect with two clients (one is the main and another readonly and verify that the readonly get the events and how the right profile state)

hhaim commented:
I've commited a workaround for event and the crash [commit](https://github.com/hhaim/trex-core/commit/8d441986082331b734d85f0af1e358dc6d8b255f)

still need to fix it. 
I've review the code, looks good. 
the tests could be better. need more intensive random test that start/stop new profiles for random duration and pool using event the changes in the profile. I would try more than 5 profile -- more in the 1000 ballpark

hhaim commented:
```
stateless_tests.stl_capture_test.STLCapture_Test.test_arp_capture
stateless_tests.stl_capture_test.STLCapture_Test.test_ping_capture
stateless_tests.stl_client_test.STLClient_Test.test_basic_cont
stateless_tests.stl_client_test.STLClient_Test.test_basic_cont_dynamic_profile
stateless_tests.stl_client_test.STLClient_Test.test_basic_multi_burst
stateless_tests.stl_client_test.STLClient_Test.test_basic_single_burst
stateless_tests.stl_client_test.STLClient_Test.test_latency_pause_resume
stateless_tests.stl_client_test.STLClient_Test.test_latency_pause_resume_dynamic_profile
stateless_tests.stl_client_test.STLClient_Test.test_pause_resume_dynamic_profile
stateless_tests.stl_client_test.STLClient_Test.test_pause_resume_update_streams
stateless_tests.stl_client_test.STLClient_Test.test_pcap_remote
stateless_tests.stl_client_test.STLClient_Test.test_push_pcap
stateless_tests.stl_client_test.STLClient_Test.test_stress_tx
stateless_tests.stl_client_test.STLClient_Test.test_tx_from_rx
stateless_tests.stl_client_test.STLClient_Test.test_update_stop_dynamic_profile
stateless_tests.stl_examples_test.STLExamples_Test.test_stl_examples
stateless_tests.stl_ipv6_test.STLIPv6_Test.test_stl_ipv6_ping
stateless_tests.stl_ipv6_test.STLIPv6_Test.test_stl_ipv6_scan6
stateless_tests.stl_rx_test.STLRX_Test.test_1_stream_many_iterations
stateless_tests.stl_rx_test.STLRX_Test.test_fcs_stream
stateless_tests.stl_rx_test.STLRX_Test.test_one_stream
stateless_tests.stl_rx_test.STLRX_Test.test_service_mode_flow_stats
stateless_tests.stl_services_test.STLServices_Test.test_arp_service
stateless_tests.trex_client_pkg_test.CTRexClientPKG_Test.test_client_python2
stateless_tests.trex_client_pkg_test.CTRexClientPKG_Test.test_client_python3

```

One example of failure 
```
stateless_tests.stl_client_test.STLClient_Test.test_latency_pause_resume (from nosetests)
Failing for the past 1 build (Since #288213 )
Took 0.22 sec.

add description

Error MessagePort 0 dest MAC is invalid and there are streams without explicit dest MAC.StacktraceTraceback (most recent call last):
  File "/usr/lib64/python2.7/unittest/case.py", line 369, in run
    testMethod()
  File "/tmp/trex-scripts/automation/regression/stateless_tests/stl_client_test.py", line 425, in test_latency_pause_resume
    assert False , '{0}'.format(e)
AssertionError: Port 0 dest MAC is invalid and there are streams without explicit dest MAC.
```
I didn't look for the reason, but it seems related to the STATE events

hhaim commented:
```

$./run_regression --stl --skip-clean --no-daemon -t test_arp_capture
Python2 test
        !!!     Running with modes: ['virt_nics', 'VM', 'loopback'], not suitable tests will be skipped.        !!!
Done instantiating TRex scenario!

test_connectivity (stateless_tests.stl_general_test.STLBasic_Test) ... 
Starting TRex... done. (0s)
Started TRex
Connecting.
Connected
Mapping ports.
Got ports mapping: {'map': {0: 1, 1: 0}, 'bi': [(0, 1)], 'unknown': []}
passed
test_arp_capture (stateless_tests.stl_capture_test.STLCapture_Test) ... ERROR
----------------------------------------------------------------------
XML: /auto/srg-sce-swinfra-usr/emb/users/hhaim/work/depot/asr1k/emb/private/hhaim/bp_sim_git/trex-core/scripts/automation/regression/reports/report_trex03_stateless.xml

-----------------------------------------------------------------------------
1) ERROR: test_arp_capture (stateless_tests.stl_capture_test.STLCapture_Test)

   Traceback (most recent call last):
    stateless_tests/stl_capture_test.py line 234 in test_arp_capture
      self.c.arp(ports = self.tx_port)
    /auto/srg-sce-swinfra-usr/emb/users/hhaim/work/depot/asr1k/emb/private/hhaim/bp_sim_git/trex-core/scripts/automation/trex_control_plane/interactive/trex/common/trex_api_annotators.py line 51 in wrap2
      ret = f(*args, **kwargs)
    /auto/srg-sce-swinfra-usr/emb/users/hhaim/work/depot/asr1k/emb/private/hhaim/bp_sim_git/trex-core/scripts/automation/trex_control_plane/interactive/trex/common/trex_client.py line 2038 in resolve
      ctx.run(arp)
    /auto/srg-sce-swinfra-usr/emb/users/hhaim/work/depot/asr1k/emb/private/hhaim/bp_sim_git/trex-core/scripts/automation/trex_control_plane/interactive/trex/common/services/trex_service_int.py line 70 in run
      self._run(services)
    /auto/srg-sce-swinfra-usr/emb/users/hhaim/work/depot/asr1k/emb/private/hhaim/bp_sim_git/trex-core/scripts/automation/trex_control_plane/interactive/trex/common/services/trex_service_int.py line 157 in _run
      self._add(services)
    /auto/srg-sce-swinfra-usr/emb/users/hhaim/work/depot/asr1k/emb/private/hhaim/bp_sim_git/trex-core/scripts/automation/trex_control_plane/interactive/trex/common/services/trex_service_int.py line 117 in _add
      self.prom_supported = port_info['prom_supported'] != 'no'
   KeyError: 'prom_supported'


-----------------------------------------------------------------------------
2 tests run in 1.6 seconds. 
1 error (1 test passed)

           /\_/\ 
          ( o.o ) 
           > ^ < 

This cat is sad, test failed.
        

*** python test failed


```

JunsubHan4rmEricsson commented:
> When you run the regression you should see the `Unhandled event 11` message
> Didn't you see this?
> 
> ```
> Unhandled event 10
> Done (10s), CPU util: 0.113771, norm_pps_per_core: 36.17mpps norm_bw_per_core: 433.24Gb/core
> Testing profile stl/udp_for_benchmarks.py, kwargs: {'packet_len': 4000}
> Unhandled event 11
> Unhandled event 11
> Unhandled event 11
> Unhandled event 11
> Unhandled event 10
> Unhandled event 10
> Unhandled event 10
> Unhandled event 10
> Done (10s), CPU util: 0.0513993, norm_pps_per_core: 30.26mpps norm_bw_per_core: 966.58Gb/core
> Testing profile stl/udp_for_benchmarks.py, kwargs: {'packet_len': 9000}
> Unhandled event 11
> Unhandled event 11
> Unhandled event 11
> Unhandled event 11
> Unhandled event 10
> Unhandled event 10
> Unhandled event 10
> Unhandled event 10
> Done (3s), CPU util: 0.034354, norm_pps_per_core: 20.23mpps norm_bw_per_core: 1220.3Gb/core
> Testing profile stl/udp_for_benchmarks.py, kwargs: {'stream_count': 10, 'packet_len': 9000}
> Unhandled event 11
> Unhandled event 11
> Unhandled event 11
> Unhandled event 11
> Unhandled event 10
> Unhandled event 10
> Unhandled event 10
> Unhandled event 10
> ```

Yes, the event is coming from the server. 
In the beginning, we added profile event not only for "STOP" and "JOB_DONE"
but also for "STARTED", "PAUSED" and "RESUMED".

For the events "STARTED", "PAUSED" and "RESUMED", from the server,
I talked to the engineer in charge of server side and decided leave them unhandled.

Two ways to solve this problem
1) Server does not send these events.
2) Client created empty functions for these. 

What do you suggest?

JunsubHan4rmEricsson commented:
1) Fix crash when reload the server (connect/disconnect a few times)
2) Handle unhandled events (for read-only clients)
3) Add more intensive randome tests for random duration (~1000 profiles)
4) Check regression failrues

I will get to it

hhaim commented:
I would start with the regression. This is something that should pass before PR the code
Even your new tests do not pass. I don’t think this is significant that regression should pass

JunsubHan4rmEricsson commented:
Working on it right now.
There was an issue with a private parameter __is_sync when syncing.
I'll keep you updated.

JunsubHan4rmEricsson commented:
> 1. Fix crash when reload the server (connect/disconnect a few times)
> 2. Handle unhandled events (for read-only clients)
> 3. Add more intensive randome tests for random duration (~1000 profiles)
> 4. Check regression failrues
> 
> I will get to it

Hi Hanoh, I was working on #4 which seems most immenent.

Changes were made and most of regression tests succeed. 
In fact, I actually ran them all myself to check after the code change. 

https://github.com/cisco-system-traffic-generator/trex-core/pull/220/commits/b6ee2d8609a76bdcb3cbc933707750890b4fbcb1

Except for "stateless_tests.stl_services_test.STLServices_Test.test_arp_service"
I am currently working on this regression to fix it.

JunsubHan4rmEricsson commented:
Hi Hanoh, 
Could you please run the regression and provide feedback?
**KeyError: 'prom_supported'** error is not there anymore.

JunsubHan4rmEricsson commented:
> I've rebased it.
> https://github.com/hhaim/trex-core/tree/bidtt4-STL_Dynamic_Profile_Client_Refined
> 
> ```
> sudo ./t-rex-64 -i
> ```
> try to connect/disconnect a few times (using quit)
> in the first time it works

Could you please share the exact commands you used to reproduce the fault?
Thanks.

JunsubHan4rmEricsson commented:
**Hi Hanoh, here's what I've done so far**
1) Handle unhandled events (for read-only clients)
https://github.com/cisco-system-traffic-generator/trex-core/pull/220/commits/57078f0860fc953df5d9dcb9dafd1ca6051187d8
2) Check regression failures
https://github.com/cisco-system-traffic-generator/trex-core/pull/220/commits/b6ee2d8609a76bdcb3cbc933707750890b4fbcb1

**What's left to be done**
1) Fix crash when reload the server (connect/disconnect a few times)
- Could you please tell me how to reproduce this error with the command list?
2) Add more intensive random tests for random duration (~1000 profiles)
- I will implement it tomorrow. You can be more specific with your needs if you want. 

In the meanwhile, could you please check the regression test?

hhaim commented:
This fix the crash. 
```diff --git a/scripts/automation/trex_control_plane/interactive/trex/common/trex_types.py b/scripts/automation/trex_control_plane/interactive/trex/common/trex_types.py
index 13444df8..3641d17c 100644
--- a/scripts/automation/trex_control_plane/interactive/trex/common/trex_types.py
+++ b/scripts/automation/trex_control_plane/interactive/trex/common/trex_types.py
@@ -46,6 +46,9 @@ class PortProfileID(object):
     def __int__(self):
         return self.port_id
 
+    def __hash__(self):
+        return hash((self.port_id, self.profile_id))
+
     def __repr__(self):
         return self.profile_name
 
```

The regression pass without your change (master). it is is related to your change. I didn't look why.

hhaim commented:
Much better now

Setup | Failed tests: | Final Result | Time (s)
-- | -- | -- | --
stateless_tests.stl_client_test.STLClient_Test.test_update_stop_dynamic_profile | FAILED | 10.709
Failure:Traceback (most recent call last):File "/usr/lib64/python2.7/unittest/case.py", line 369, in runtestMethod()File "/tmp/trex-scripts/automation/regression/stateless_tests/stl_client_test.py", line 852, in test_update_stop_dynamic_profileself.check_error_in_percentage_dynamic_profile(golden*2)File "/tmp/trex-scripts/automation/regression/stateless_tests/stl_client_test.py", line 721, in check_error_in_percentage_dynamic_profileassert get_error_in_percentage(golden, param) < 0.05, 'golden: %s, got: %s' % (golden, param)AssertionError: golden: 2000, got: 3550.63012695
stateless_tests.stl_core_id_pinning_test.CoreID_Test.test_core_id_pinning_several_cores | FAILED | 0.665
Failure:Traceback (most recent call last):File "/usr/lib64/python2.7/unittest/case.py", line 369, in runtestMethod()File "/tmp/trex-scripts/automation/regression/stateless_tests/stl_core_id_pinning_test.py", line 140, in test_core_id_pinning_several_coresassert sorted(used_cores) == sorted(cores_list), 'Expected working cores: %s, got: %s' % (cores_list, used_cores)AssertionError: Expected working cores: [1, 4, 7], got: [0, 1, 2, 3, 4, 5, 6, 7]
stateless_tests.stl_core_id_pinning_test.CoreID_Test.test_core_id_pinning_several_streams | FAILED | 0.662
Failure:Traceback (most recent call last):File "/usr/lib64/python2.7/unittest/case.py", line 369, in runtestMethod()File "/tmp/trex-scripts/automation/regression/stateless_tests/stl_core_id_pinning_test.py", line 158, in test_core_id_pinning_several_streamsassert used_cores == [core_id], 'Only core %s is expected to be active, got: %s' % (core_id, used_cores)AssertionError: Only core 7 is expected to be active, got: [0, 1, 2, 3, 4, 5, 6, 7]
stateless_tests.stl_core_id_pinning_test.CoreID_Test.test_core_id_pinning_single | FAILED | 0.665
Failure:Traceback (most recent call last):File "/usr/lib64/python2.7/unittest/case.py", line 369, in runtestMethod()File "/tmp/trex-scripts/automation/regression/stateless_tests/stl_core_id_pinning_test.py", line 124, in test_core_id_pinning_singleassert used_cores == [i], 'Only core %s is expected to be active, but we got %s active' % (i, used_cores)AssertionError: Only core 0 is expected to be active, but we got [0, 1, 2, 3, 4, 5, 6, 7] active
stateless_tests.stl_core_id_pinning_test.CoreID_Test.test_core_id_pinning_single_burst_loop | FAILED | 0.658
Failure:Traceback (most recent call last):File "/usr/lib64/python2.7/unittest/case.py", line 369, in runtestMethod()File "/tmp/trex-scripts/automation/regression/stateless_tests/stl_core_id_pinning_test.py", line 170, in test_core_id_pinning_single_burst_loopassert used_cores == [core_id], 'Only core %s is expected to be active, got: %s' % (core_id, used_cores)AssertionError: Only core 7 is expected to be active, got: [0, 1, 2, 3, 4, 5, 6, 7]
stateless_tests.stl_client_test.STLClient_Test.test_latency_pause_resume_dynamic_profile | FAILED | 2.042
Failure:Traceback (most recent call last):File "/usr/lib64/python2.7/unittest/case.py", line 369, in runtestMethod()File "/tmp/trex-scripts/automation/regression/stateless_tests/stl_client_test.py", line 913, in test_latency_pause_resume_dynamic_profileassert False , '{0}'.format(e)AssertionError: Port 3 : *** For payload rules: Can't have two streams with same pg_id, or same stream on more than one port
stateless_tests.stl_client_test.STLClient_Test.test_update_stop_dynamic_profile | FAILED | 10.455
Failure:Traceback (most recent call last):File "/usr/lib64/python2.7/unittest/case.py", line 369, in runtestMethod()File "/tmp/trex-scripts/automation/regression/stateless_tests/stl_client_test.py", line 852, in test_update_stop_dynamic_profileself.check_error_in_percentage_dynamic_profile(golden*2)File "/tmp/trex-scripts/automation/regression/stateless_tests/stl_client_test.py", line 721, in check_error_in_percentage_dynamic_profileassert get_error_in_percentage(golden, param) < 0.05, 'golden: %s, got: %s' % (golden, param)AssertionError: golden: 2000, got: 2861.82177734
stateless_tests.stl_core_id_pinning_test.CoreID_Test.test_core_id_negative | ERROR | 0.553
Error:Traceback (most recent call last):File "/usr/lib64/python2.7/unittest/case.py", line 360, in runself.setUp()File "/tmp/trex-scripts/automation/regression/stateless_tests/stl_core_id_pinning_test.py", line 25, in setUpassert self.num_cores == len(self.tx_port_cores), '%s != %s' % (self.num_cores, len(self.tx_port_cores))AssertionError: 7 != 14
stateless_tests.stl_core_id_pinning_test.CoreID_Test.test_core_id_pinning_several_cores | ERROR | 0.548
Error:Traceback (most recent call last):File "/usr/lib64/python2.7/unittest/case.py", line 360, in runself.setUp()File "/tmp/trex-scripts/automation/regression/stateless_tests/stl_core_id_pinning_test.py", line 25, in setUpassert self.num_cores == len(self.tx_port_cores), '%s != %s' % (self.num_cores, len(self.tx_port_cores))AssertionError: 7 != 14
stateless_tests.stl_core_id_pinning_test.CoreID_Test.test_core_id_pinning_several_streams | ERROR | 0.554
Error:Traceback (most recent call last):File "/usr/lib64/python2.7/unittest/case.py", line 360, in runself.setUp()File "/tmp/trex-scripts/automation/regression/stateless_tests/stl_core_id_pinning_test.py", line 25, in setUpassert self.num_cores == len(self.tx_port_cores), '%s != %s' % (self.num_cores, len(self.tx_port_cores))AssertionError: 7 != 14
stateless_tests.stl_core_id_pinning_test.CoreID_Test.test_core_id_pinning_single | ERROR | 0.569
Error:Traceback (most recent call last):File "/usr/lib64/python2.7/unittest/case.py", line 360, in runself.setUp()File "/tmp/trex-scripts/automation/regression/stateless_tests/stl_core_id_pinning_test.py", line 25, in setUpassert self.num_cores == len(self.tx_port_cores), '%s != %s' % (self.num_cores, len(self.tx_port_cores))AssertionError: 7 != 14
stateless_tests.stl_core_id_pinning_test.CoreID_Test.test_core_id_pinning_single_burst_loop | ERROR | 0.563
Error:Traceback (most recent call last):File "/usr/lib64/python2.7/unittest/case.py", line 360, in runself.setUp()File "/tmp/trex-scripts/automation/regression/stateless_tests/stl_core_id_pinning_test.py", line 25, in setUpassert self.num_cores == len(self.tx_port_cores), '%s != %s' % (self.num_cores, len(self.tx_port_cores))AssertionError: 7 != 14
stateless_tests.stl_client_test.STLClient_Test.test_pause_resume_dynamic_profile | FAILED | 10.582
Failure:Traceback (most recent call last):File "/usr/lib64/python2.7/unittest/case.py", line 369, in runtestMethod()File "/tmp/trex-scripts/automation/regression/stateless_tests/stl_client_test.py", line 804, in test_pause_resume_dynamic_profileself.check_error_in_percentage_dynamic_profile(golden/2)File "/tmp/trex-scripts/automation/regression/stateless_tests/stl_client_test.py", line 721, in check_error_in_percentage_dynamic_profileassert get_error_in_percentage(golden, param) < 0.05, 'golden: %s, got: %s' % (golden, param)AssertionError: golden: 500, got: 647.227416992

JunsubHan4rmEricsson commented:
> **Hi Hanoh, here's what I've done so far**
> 
> 1. Handle unhandled events (for read-only clients)
>    [57078f0](https://github.com/cisco-system-traffic-generator/trex-core/commit/57078f0860fc953df5d9dcb9dafd1ca6051187d8)
> 2. Check regression failures
>    [b6ee2d8](https://github.com/cisco-system-traffic-generator/trex-core/commit/b6ee2d8609a76bdcb3cbc933707750890b4fbcb1)
>    ****KeyError: 'prom_supported**' error is not there anymore.
> 
> **What's left to be done**
> 
> 1. Fix crash when reload the server (connect/disconnect a few times)
> 
> * Could you please tell me how to reproduce this error with the command list?
> 
> 1. Add more intensive random tests for random duration (~1000 profiles)
> 
> * I will implement it tomorrow. You can be more specific with your needs if you want.
> 
> **In the meanwhile, could you please run the regression test?**
> 
> PS: Even after working on the regression failures, there seems to be an error with test_arp_services (assertion error, "assert rec.dst_mac == arps[0].get_record().dst_mac")
> However, I ran it with the legacy version trex and the same fault takes place.
> Therefore, it is not caused by our source changes.

Okay, then
Crash, regression failures and unhandled events are all taken care of. 
What do you suggest that I do now.

Just add more profile test cases?
When will this be merged?

hhaim commented:
Regression is still not OK. It is much better but not OK.
Even your new tests fails. Start with core pinning test.
Have a look into the error I’ve sent 
We will try to give access to our regression from remote branch.

JunsubHan4rmEricsson commented:
I checked. 
When I run the cases in my environment. 
No errors or failures take place.

hhaim commented:
Can you say what are the platforms type you are testing this?

JunsubHan4rmEricsson commented:
**cat /etc/os-release** 
NAME="Ubuntu"
VERSION="18.04.1 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.1 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic

**lsb_release -a**
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.1 LTS
Release:	18.04
Codename:	bionic

**uname -r**
4.15.0-42-generic

hhaim commented:
I mean more from  interfaces types/number of cores/ baremeta and not baremetal etc
the faliures happen in our trex-08/09 

see here
[setups](https://github.com/cisco-system-traffic-generator/trex-core/wiki#setups)

it is with XL710/X710, bare metal with c==14/3 per interfaces. 

running this tests manualy does not fails, I'm looking into this

JunsubHan4rmEricsson commented:
**cat /etc/trex_cfg.yaml** 
- port_limit: 2
  version: 2
  c: 2
  prefix: default
  interfaces: ['3b:0a.1', '3b:0a.2']
  zmq_pub_port: 4500
  zmq_rpc_port: 4501
  limit_memory: 2048
  rx_desc : 4096
  port_info:
      - ip: 1.1.1.1
        default_gw: 2.2.2.2
      - ip: 2.2.2.2
        default_gw: 1.1.1.1
#    - dest_mac: 02:df:4b:5f:04:62
#    - dest_mac: 02:df:4b:5f:05:62
  platform:
    master_thread_id: 2
    latency_thread_id: 4
    dual_if:
      - socket: 0
        threads: [6,8]

**./dpdk_nic_bind.py -s**
Network devices using DPDK-compatible driver
============================================
0000:3b:02.0 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:02.1 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:02.2 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:02.3 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:02.4 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:02.5 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:02.6 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:02.7 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:03.0 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:03.1 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:03.2 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:03.3 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:03.4 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:03.5 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:03.6 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:03.7 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0a.0 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
**0000:3b:0a.1 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0a.2 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf**
0000:3b:0a.3 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0a.4 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0a.5 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0a.6 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0a.7 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0b.0 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0b.1 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0b.2 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0b.3 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0b.4 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0b.5 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0b.6 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0b.7 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0c.0 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0c.1 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0c.2 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0c.3 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0c.4 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0c.5 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0c.6 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0c.7 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0d.0 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0d.1 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0d.2 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0d.3 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0d.4 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0d.5 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0d.6 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0d.7 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:5e:10.0 '82599 Ethernet Controller Virtual Function' drv=vfio-pci unused=ixgbevf
0000:5e:10.2 '82599 Ethernet Controller Virtual Function' drv=vfio-pci unused=ixgbevf
0000:5e:10.4 '82599 Ethernet Controller Virtual Function' drv=vfio-pci unused=ixgbevf
0000:5e:10.6 '82599 Ethernet Controller Virtual Function' drv=vfio-pci unused=ixgbevf

Network devices using kernel driver
===================================
0000:01:00.0 'I350 Gigabit Network Connection' if=eno3 drv=igb unused=vfio-pci *Active*
0000:01:00.1 'I350 Gigabit Network Connection' if=eno4 drv=igb unused=vfio-pci 
0000:19:00.0 '82599ES 10-Gigabit SFI/SFP+ Network Connection' if=eno1 drv=ixgbe unused=vfio-pci *Active*
0000:19:00.1 '82599ES 10-Gigabit SFI/SFP+ Network Connection' if=eno2 drv=ixgbe unused=vfio-pci 
0000:3b:00.0 'Ethernet Controller XXV710 for 25GbE SFP28' if=enp59s0f0 drv=i40e unused=vfio-pci 
0000:3b:00.1 'Ethernet Controller XXV710 for 25GbE SFP28' if=enp59s0f1 drv=i40e unused=vfio-pci 
0000:5e:00.0 'Ethernet 10G 2P X520 Adapter' if=enp94s0f0 drv=ixgbe unused=vfio-pci 
0000:5e:00.1 'Ethernet 10G 2P X520 Adapter' if=enp94s0f1 drv=ixgbe unused=vfio-pci 
0000:5e:11.0 '82599 Ethernet Controller Virtual Function' if=enp94s17 drv=ixgbevf unused=vfio-pci 
0000:5e:11.2 '82599 Ethernet Controller Virtual Function' if=enp94s17f2 drv=ixgbevf unused=vfio-pci 
0000:5e:11.4 '82599 Ethernet Controller Virtual Function' if=enp94s17f4 drv=ixgbevf unused=vfio-pci 
0000:5e:11.6 '82599 Ethernet Controller Virtual Function' if=enp94s17f6 drv=ixgbevf unused=vfio-pci 
0000:5e:12.0 '82599 Ethernet Controller Virtual Function' if=enp94s18 drv=ixgbevf unused=vfio-pci 
0000:5e:12.2 '82599 Ethernet Controller Virtual Function' if=enp94s18f2 drv=ixgbevf unused=vfio-pci 
0000:5e:12.4 '82599 Ethernet Controller Virtual Function' if=enp94s18f4 drv=ixgbevf unused=vfio-pci 
0000:5e:12.6 '82599 Ethernet Controller Virtual Function' if=enp94s18f6 drv=ixgbevf unused=vfio-pci 
0000:5e:13.0 '82599 Ethernet Controller Virtual Function' if=enp94s19 drv=ixgbevf unused=vfio-pci 
0000:5e:13.2 '82599 Ethernet Controller Virtual Function' if=enp94s19f2 drv=ixgbevf unused=vfio-pci 
0000:5e:13.4 '82599 Ethernet Controller Virtual Function' if=enp94s19f4 drv=ixgbevf unused=vfio-pci 
0000:5e:13.6 '82599 Ethernet Controller Virtual Function' if=enp94s19f6 drv=ixgbevf unused=vfio-pci 

Other network devices
=====================
<none>
root@seliics03192:~/ehajuns/opt/etrex#

hhaim commented:
I found a lead to the issues. It seems your tests *_dynamic_profile run before the core pinning and make them fails. Removing the your tests make the regression pass on trex-08. 
Will check it on all platforms.

hhaim commented:
You should have a baremetal XL710 platform. running the full regression will fail

JunsubHan4rmEricsson commented:
These are the interfaces used for regression. 
It works just fine. 

0000:3b:0a.1 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf
0000:3b:0a.2 'XL710/X710 Virtual Function' drv=vfio-pci unused=i40evf

JunsubHan4rmEricsson commented:
> I found a lead to the issues. It seems your tests *_dynamic_profile run before the core pinning and make them fails. Removing the your tests make the regression pass on trex-08.
> Will check it on all platforms.

Yes let me know how it goes. 
I will check tomorrow based your feedback

hhaim commented:
One more issue, you duplicate the sync function from common but you miss the block.
Without that linux_stack won't work 

```
     def sync(self):
         params = {"port_id": self.port_id,
-                  "profile_id": ALL_PROFILE_ID}
+                  "profile_id": ALL_PROFILE_ID,'block': False}
 
         rc = self.transmit("get_port_status", params)
         if rc.bad():
```

hhaim commented:
1. Run the regression on a bare metal setup (see our trex-08/09) and not on VF as the regression is limited on VF/Virtual platforms 
We will work on a way for you to run our regression

* XL710 PF (not vf)
* c=7 (number of cores per dual) 
* bare metal 
* loopback the ports

2. Enable the dynmaic_profiles tests back and find why it is not stable with 1 as a setup  

3. Add more stress for dynamic profile 

test args:
  max_active_profile: how many active profiles, e.g. 1000
  profile_name: e.g. imix.py
  test_duration_sec: how many sec to run the test
  each_profile_max_duration: the maximum duration to run each profile 

```
profiles = []

while True :

  delay_random(50-100msec)
  
  event_type = choose_event() (could be (add , remove)
  
  if duration > test_duration 
      break;
  
  switch event:

     add: 
        id++
        active++
        duration_sec = random((10..test_duration/2))
        profiles.append(id)
        if active<1000:
           add_imix_profile(profile.id,duration_sec)

     remove: 
       # choose random profile profiles list and remove it
         rp = choose_random_profile()
         c.remove_profile(rp)
         active--

  # check the events of profile that has stopped due to duration 
  event_from_client = pool_events_from_client()
  if event_from_client == profile_stopped :
     c.remove_profile(event_from_client.profile_name)
     active--
   
remove_all_profiles()

```

JunsubHan4rmEricsson commented:
> One more issue, you duplicate the sync function from common but you miss the block.
> Without that linux_stack won't work
> 
> ```
>      def sync(self):
>          params = {"port_id": self.port_id,
> -                  "profile_id": ALL_PROFILE_ID}
> +                  "profile_id": ALL_PROFILE_ID,'block': False}
>  
>          rc = self.transmit("get_port_status", params)
>          if rc.bad():
> ```

Thanks.
I pushed the changes.
https://github.com/cisco-system-traffic-generator/trex-core/pull/220/commits/958aab3db93bf2c3884ceeb195ab945d7888e61b

JunsubHan4rmEricsson commented:
> 1. Run the regression on a bare metal setup (see our trex-08/09) and not on VF as the regression is limited on VF/Virtual platforms
>    We will work on a way for you to run our regression
> 
> * XL710 PF (not vf)
> * c=7 (number of cores per dual)
> * bare metal
> * loopback the ports
> 
> 1. Enable the dynmaic_profiles tests back and find why it is not stable with 1 as a setup
> 2. Add more stress for dynamic profile
> 
> test args:
> max_active_profile: how many active profiles, e.g. 1000
> profile_name: e.g. imix.py
> test_duration_sec: how many sec to run the test
> each_profile_max_duration: the maximum duration to run each profile
> 
> ```
> profiles = []
> 
> while True :
> 
>   delay_random(50-100msec)
>   
>   event_type = choose_event() (could be (add , remove)
>   
>   if duration > test_duration 
>       break;
>   
>   switch event:
> 
>      add: 
>         id++
>         active++
>         duration_sec = random((10..test_duration/2))
>         profiles.append(id)
>         if active<1000:
>            add_imix_profile(profile.id,duration_sec)
> 
>      remove: 
>        # choose random profile profiles list and remove it
>          rp = choose_random_profile()
>          c.remove_profile(rp)
>          active--
> 
>   # check the events of profile that has stopped due to duration 
>   event_from_client = pool_events_from_client()
>   if event_from_client == profile_stopped :
>      c.remove_profile(event_from_client.profile_name)
>      active--
>    
> remove_all_profiles()
> ```

Like I've said all tests including the profiles run just fine on our environment. 
However, I do get your point about setting the environment for regression.
It's not so easy for us to set up the environment exactly like that however. 

So #1 and #2 are not so easy for us to check.
You're going to have to help us out a little. 

#3 however, I can start working on that.

hhaim commented:
The code is already in the main branch, so it is better to close this thread.
Regarding the setup. You have the hardware it is just running it in a different way (PF vs VF). In VF case we skip many tests. It is important to debug this feature with the right configuration

JunsubHan4rmEricsson commented:
> The code is already in the main branch, so it is better to close this thread.
> Regarding the setup. You have the hardware it is just running it in a different way (PF vs VF). In VF case we skip many tests. It is important to debug this feature with the right configuration

Thanks.
I will talk to my team members about the detailed tests
